### PR TITLE
[lld][ELF] Support relax R_LARCH_ALIGN

### PR DESCRIPTION
**Pull request by MQ-mengqing**
Refer to commit 6611d58f5bbc ("Relax R_RISCV_ALIGN"), we can relax R_LARCH_ALIGN by same way. Reuse `SymbolAnchor`, `RISCVRelaxAux` and `initSymbolAnchors` to simplify codes. As `riscvFinalizeRelax` is an arch-specific function, put it override on `TargetInfo::finalizeRelax`, so that LoongArch can override it, too.

The flow of relax R_LARCH_ALIGN is almost consistent with RISCV. The difference is that LoongArch only has 4-bytes NOP and all executable insn is 4-bytes aligned. So LoongArch not need rewrite NOP sequence. Alignment maxBytesEmit parameter is supported in psABI v2.30.

### CODE REVIEW

**Comment by MQ-mengqing** [comment]
Add: @SixWeining @MaskRay @xen0n

**Review by SixWeining — COMMENTED** [review]
Basically looks good to me. Thanks.
@MaskRay What do you think about this change? If looks OK, do you mind merging this before 18.x is created or cherry-picking later? As gcc/binutils have began generating relax relocations for about 1 year, lld needs this PR so that it can link fundamental system objects (such as `Scrt1.o`).

File: lld/test/ELF/loongarch-relax-align.s [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+# CHECKR-NEXT:         {{0*}}04: R_LARCH_RELAX      *ABS*
+# CHECKR-NEXT:  8:     nop
+# CHECKR-NEXT:         {{0*}}08: R_LARCH_ALIGN      .Lalign_symbol+0x4
+# CHECKR-NEXT:  c:     nop
+# CHECKR-NEXT: 10:     nop
+# CHECKR-NEXT: 14:     break 6
+
+.macro .fake_p2align_4 max=0
+  .ifdef old
+    .ifeq \max!=0
```
[the remark this answers is not in the document]
Comment (MQ-mengqing) [reply lld/test/ELF/loongarch-relax-align.s:85]:
```
.ifeq absolute expression
Assembles the following section of code if the argument is zero.
```
It's counter-intuitive. I may miss `.if` in `Using as`, it should be more intuitive. I'll correct it.

**Comment by xry111** [comment]
> Basically looks good to me. Thanks. @MaskRay What do you think about this change? If looks OK, do you mind merging this before 18.x is created or cherry-picking later? As gcc/binutils have began generating relax relocations for about 1 year, lld needs this PR so that it can link fundamental system objects (such as `Scrt1.o`).

IMO at least supporting R_LARCH_RELAX should be considered a bug fix and it should be in 18.x, or we are basically shipping "unusable" things.

**Comment by xen0n** [comment]
Sorry for the late reply, but I agree with cherry-picking this change into LLVM 18, at least the `R_LARCH_ALIGN` bits. Because while linker relaxation is strictly a performance optimization and can safely be omitted without creating issues besides slower code, the unfortunate fact is that `R_LARCH_ALIGN` is already being used to implement alignment directives (so it's already popping out in the wild), and correctly handling of this reloc apparently requires implementing enough of relaxation mechanism.

File: lld/ELF/Arch/LoongArch.cpp [hunk 2]
```diff
@@ -659,6 +664,253 @@ void LoongArch::relocate(uint8_t *loc, const Relocation &rel,
...
+
+static bool relax(InputSection &sec) {
+  const uint64_t secAddr = sec.getVA();
+  auto &aux = *sec.relaxAux;
+  bool changed = false;
+  ArrayRef<SymbolAnchor> sa = ArrayRef(aux.anchors);
+  uint64_t delta = 0;
+
+  std::fill_n(aux.relocTypes.get(), sec.relocs().size(), R_LARCH_NONE);
+  aux.writes.clear();
```
Comment (MaskRay) [inline lld/ELF/Arch/LoongArch.cpp:730]: In RISCV.cpp, I have defined `relocs` as a variable to hold `src.relocs()`

File: lld/ELF/InputSection.h [hunk 3]
```diff
@@ -225,9 +241,9 @@ class InputSectionBase : public SectionBase {
     // basic blocks.
     JumpInstrMod *jumpInstrMod = nullptr;
 
-    // Auxiliary information for RISC-V linker relaxation. RISC-V does not use
-    // jumpInstrMod.
-    RISCVRelaxAux *relaxAux;
+    // Auxiliary information for RISC-V and LoongArch linker relaxation.
+    // RISC-V and LoongArch does not use jumpInstrMod.
```
Comment (MaskRay) [inline lld/ELF/InputSection.h:245]: They do not use `jumpInstrMod`

File: lld/ELF/Arch/LoongArch.cpp [hunk 4]
```diff
@@ -659,6 +664,253 @@ void LoongArch::relocate(uint8_t *loc, const Relocation &rel,
...
+// shrinkage may reduce displacement and make more relocations eligible for
+// relaxation. Code shrinkage may increase displacement to a call/load/store
+// target at a higher fixed address, invalidating an earlier relaxation. Any
+// change in section sizes can have cascading effect and require another
+// relaxation pass.
+bool LoongArch::relaxOnce(int pass) const {
+  if (config->relocatable)
+    return false;
+
+  // TODO Relax multiple times.
```
Comment (MaskRay) [inline lld/ELF/Arch/LoongArch.cpp:809]: Why is this here?

File: lld/ELF/Arch/LoongArch.cpp [hunk 5]
```diff
@@ -659,6 +664,253 @@ void LoongArch::relocate(uint8_t *loc, const Relocation &rel,
...
+  const Symbol &sym = *r.sym;
+  const uint32_t pca = read64le(sec.content().data() + r.offset);
+  const uint32_t ld = read64le(sec.content().data() + rl.offset);
+  const uint32_t rd = pca & 0x1f;
+  const uint32_t op_addi = config->is64 ? ADDI_D : ADDI_W;
+  const uint32_t op_ld = config->is64 ? LD_D : LD_W;
+  const uint64_t dest = sym.getVA() + r.addend;
+  const int64_t displace = dest - loc;
+
+  if (!sym.isLocal())
```
Comment (MaskRay) [inline lld/ELF/Arch/LoongArch.cpp:702]: This seems strange. `isPreemptible decides whether a GOT can be optimized.

File: lld/test/ELF/loongarch-relax-align.s [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+# RELAX-DAG: {{0*}}1002c l .text  {{0*}}08 .L1
+# RELAX-DAG: {{0*}}10030 l .text  {{0*}}04 .L2
+# RELAX-DAG: {{0*}}20000 l .text2 {{0*}}14 .Ltext2_start
+# NOREL-DAG: {{0*}}10000 l .text  {{0*}}44 .Ltext_start
+# NOREL-DAG: {{0*}}10038 l .text  {{0*}}0c .L1
+# NOREL-DAG: {{0*}}10040 l .text  {{0*}}04 .L2
+# NOREL-DAG: {{0*}}20000 l .text2 {{0*}}14 .Ltext2_start
+
+# CHECK:      <.Ltext_start>:
+# CHECK-NEXT:    10000:      break 1
```
Comment (MaskRay) [inline lld/test/ELF/loongarch-relax-align.s:35]: For newer tests, omit addresses then they are insignificant

File: lld/ELF/Arch/LoongArch.cpp (hunk 4 above)
Comment (MQ-mengqing) [reply lld/ELF/Arch/LoongArch.cpp:809]: When I did some edge tests, it triggers "relaxation not converged". In the following example, <1> and <2> only one can be relaxed and cause `pass` increasing. So I disable multiple times relax for stability.

```
.global sym1
.global sym2
.type sym1, @function
.type sym2, @function

.section ".sec1", "ax"
sym1:
la.pcrel $t0, .L2 - 8  # <1>
.L1:
break 0
.size sym1, . - sym1

.section ".sec2", "ax"
sym2:
.p2align 3
.p2align 3
la.pcrel $t0, .L1 - 8  # <2>
.L2:
break 0
.size sym2, . - sym2

$ ./bin/llvm-mc --filetype=obj --triple=loongarch64 --mattr=+relax tmp.s -o tmp.o
$ ./bin/ld.lld tmp.o -e 0 --section-start=.sec1=0x000000 --section-start=.sec2=0x200000
$ ld.lld: error: relaxation not converged
```

Comment (xry111) [reply lld/ELF/Arch/LoongArch.cpp:809]: Interestingly, ld.bfd relaxed neither.

Comment (MQ-mengqing) [reply lld/ELF/Arch/LoongArch.cpp:809]: ld.bfd relax R_LARCH_ALIGN at the final pass. So it not conflict in this test. But it also has some shortages. Some symbol values will change due to final relax R_LARCH_ALIGN, so that the previous relax may fail.
```
.text
.p2align 4
.irp t,0,1,2,3,4,5,6,7,8,9,a,b,c,d,e,f
  .p2align 10
.endr
.p2align 6
.p2align 4
la.pcrel $s9, .L1

.section .text2, "ax"
.p2align 4
la.pcrel $r0, .L1
.data
.L1: .word 0

$ ./gas/as-new 1.s -o 1.o
$ ./ld/ld-new 1.o -e 0 -Ttext=0x100000 -Tdata=0x300000
./ld/ld-new: 1.o: relocation R_LARCH_PCREL20_S2 overflow 0x200000
```
In my idea, I think the relax method MaskRay supported for RISCV is better. I keep here just one time relaxation because I'm also afraid that I don't know enough this method. In this relaxation all R_LARCH_ALIGN will be relaxed, which is expected.

**Comment by MaskRay** [comment]
For release/18.x, I think relaxing just `R_*_ALIGN` is safer. We need more time to think about GOT relaxation. Relaxing once may not converge when other features like RELR are involved. Can you split the patch?

**Comment by SixWeining** [comment]
> For release/18.x, I think relaxing just `R_*_ALIGN` is safer. We need more time to think about GOT relaxation. Relaxing once may not converge when other features like RELR are involved. Can you split the patch?

I agree about the spliting. In addition, I think lld-18 needs to process `R_LARCH_{ADD,SUB}_ULEB128` as well because there may be such relocs in `.gcc_except_table` section when `A` and `B` are separated by alignment directives. (Both gcc and clang emit alignment directive before `for` loop).

**Comment by MaskRay** [comment]
> > For release/18.x, I think relaxing just `R_*_ALIGN` is safer. We need more time to think about GOT relaxation. Relaxing once may not converge when other features like RELR are involved. Can you split the patch?
> 
> I agree about the spliting. In addition, I think lld-18 needs to process `R_LARCH_{ADD,SUB}_ULEB128` as well because there may be such relocs in `.gcc_except_table` section when `A` and `B` are separated by alignment directives. (Both gcc and clang emit alignment directive before `for` loop).

Well, if it can't, it's largely due to GCC/binutils moving too fast on relaxation :( (which I warned long time ago).

**Comment by MQ-mengqing** [comment]
Split it and support relax R_LARCH_ALIGN only. I've force pushed it and updated the title and commit info. Please review. Thanks!

**Comment by SixWeining** [comment]
> > > For release/18.x, I think relaxing just `R_*_ALIGN` is safer. We need more time to think about GOT relaxation. Relaxing once may not converge when other features like RELR are involved. Can you split the patch?
> > 
> > 
> > I agree about the spliting. In addition, I think lld-18 needs to process `R_LARCH_{ADD,SUB}_ULEB128` as well because there may be such relocs in `.gcc_except_table` section when `A` and `B` are separated by alignment directives. (Both gcc and clang emit alignment directive before `for` loop).
> 
> Well, if it can't, it's largely due to GCC/binutils moving too fast on relaxation :( (which I warned long time ago).

Patch for uleb128 is relative easy (already done locally) and will submitted after this PR is merged.

File: lld/ELF/Arch/LoongArch.cpp [hunk 7]
```diff
@@ -659,6 +663,156 @@ void LoongArch::relocate(uint8_t *loc, const Relocation &rel,
...
+      const uint64_t maxBytes = addend >> 8;
+      const uint64_t off = loc & (align - 1);
+      const uint64_t curBytes = off == 0 ? 0 : align - off;
+      // All bytes beyond the alignment boundary should be removed.
+      // If emit bytes more than max bytes to emit, remove all.
+      if (maxBytes != 0 && curBytes > maxBytes)
+        remove = allBytes;
+      else
+        remove = allBytes - curBytes;
+      assert(static_cast<int32_t>(remove) >= 0 &&
```
Comment (MaskRay) [inline lld/ELF/Arch/LoongArch.cpp:694]: RISC-V has changed this to an `errorOrWarn`

File: lld/ELF/Arch/LoongArch.cpp [hunk 8]
```diff
@@ -659,6 +663,156 @@ void LoongArch::relocate(uint8_t *loc, const Relocation &rel,
...
+    if (!(osec->flags & SHF_EXECINSTR))
+      continue;
+    for (InputSection *sec : getInputSections(*osec, storage))
+      changed |= relax(*sec);
+  }
+  return changed;
+}
+
+void LoongArch::finalizeRelax(int passes) const {
+  SmallVector<InputSection *, 0> storage;
```
Comment (MaskRay) [inline lld/ELF/Arch/LoongArch.cpp:755]: Add `log("relaxation passes: " + Twine(passes));` like RISCV.cpp does

File: lld/ELF/Arch/LoongArch.cpp [hunk 9]
```diff
@@ -659,6 +663,156 @@ void LoongArch::relocate(uint8_t *loc, const Relocation &rel,
...
+        if (remove == 0 && aux.relocTypes[i] == R_LARCH_NONE)
+          continue;
+
+        // Copy from last location to the current relocated location.
+        const Relocation &r = rels[i];
+        uint64_t size = r.offset - offset;
+
+        // The rel.type may be fixed to R_LARCH_RELAX and the instruction
+        // will be deleted, then cause size<0
+        if (remove == 0 && (int64_t)size < 0) {
```
Comment (MaskRay) [inline lld/ELF/Arch/LoongArch.cpp:788]: This whole `if` is unneeded? No test fail.

File: lld/ELF/Arch/LoongArch.cpp [hunk 10]
```diff
@@ -659,6 +663,156 @@ void LoongArch::relocate(uint8_t *loc, const Relocation &rel,
...
+          assert(aux.relocTypes[i] == R_LARCH_RELAX && "Unexpected size");
+          continue;
+        }
+
+        memcpy(p, old.data() + offset, size);
```
Comment (MaskRay) [inline lld/ELF/Arch/LoongArch.cpp:793]: RISCV.cpp has just one memcpy and it is immediately after `uint64_t size = r.offset - offset;`

File: lld/test/ELF/loongarch-relax-align.s [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+  break 4
+  .fake_p2align_4 8
+## +0x18: Emit 2 nops <= 8 bytes.
+
+## Compensate
+.ifdef old
+  nop; nop
+.endif
+
+## +0x20: Test symbol value and symbol size can be handled.
```
Comment (MaskRay) [inline lld/test/ELF/loongarch-relax-align.s:106]: The symbol value test may be weak. `riscv-relax-align.s` carefully covers many cases.

**Comment by xry111** [comment]
I'd suggest make some tests to see if `--gc-section` works with relaxation.  In the BFD linker we found `--gc-section` is somehow broken now with relaxation :(

**Comment by MaskRay** [comment]
> I'd suggest make some tests to see if `--gc-section` works with relaxation. In the BFD linker we found `--gc-section` is somehow broken now with relaxation :(

With lld's structure, a `--gc-sections` test is not necessary. I'd add one test like `lld/test/ELF/riscv-relax-emit-relocs.s`

File: lld/ELF/Arch/LoongArch.cpp (hunk 10 above)
Comment (MQ-mengqing) [reply lld/ELF/Arch/LoongArch.cpp:793]: The first memcpy is copy from the last offset to current reloc offset. The second memcpy is copy from final offset the to the section end.

File: lld/test/ELF/loongarch-relax-align.s (hunk 11 above)
Comment (MQ-mengqing) [reply lld/test/ELF/loongarch-relax-align.s:106]: I think it is enough for check symbol value. `.Ltext_start` checks value is not changed but size is changed.  `.L1` checks value is changed and size is changed.  `.L2` checks value is changed but size is not changed.

**Review by MaskRay — APPROVED** [review]

**Comment by SixWeining** [comment]
> > I'd suggest make some tests to see if `--gc-section` works with relaxation. In the BFD linker we found `--gc-section` is somehow broken now with relaxation :(
> 
> With lld's structure, a `--gc-sections` test is not necessary. I'd add one test like `lld/test/ELF/riscv-relax-emit-relocs.s`

But `--emit-relocs` doesn't work because R_LARCH_RELAX is ignored by `return R_NONE;` in `LoongArch::getRelExpr`.

**Comment by MaskRay** [comment]
> > > I'd suggest make some tests to see if `--gc-section` works with relaxation. In the BFD linker we found `--gc-section` is somehow broken now with relaxation :(
> > 
> > 
> > With lld's structure, a `--gc-sections` test is not necessary. I'd add one test like `lld/test/ELF/riscv-relax-emit-relocs.s`
> 
> But `--emit-relocs` doesn't work because R_LARCH_RELAX is ignored by `return R_NONE;` in `LoongArch::getRelExpr`.

Just test the current behavior, so that a future change supporting `R_LARCH_RELAX` can demonstrate the effect.

**Review by SixWeining — APPROVED** [review]

**Comment by SixWeining** [comment]
I have created a backport issue #80789 for this PR.
